src/libs/Queue/Queue.ts

```diff
@@ -0,0 +1,68 @@
+import type Queue from './QueueType';
+
+// Function to create a new queue
+function createQueue<T>(processItem: (item: T) => Promise<void>): Queue<T> {
+    // Array to hold the elements of the queue
+    const elements: T[] = [];
+    let isProcessing = false;
+
+    // Function to remove an item from the front of the queue
+    function dequeue(): T | undefined {
+        return elements.shift();
+    }
+
+    // Function to check if the queue is empty
+    function isEmpty(): boolean {
+        return elements.length === 0;
+    }
+
+    // Initiates the processing of items in the queue.
+    // Continues to dequeue and process items as long as the queue is not empty.
+    // Sets the `isProcessing` flag to true at the start and resets it to false once all items have been processed.
+    function run(): Promise<void> {
+        return new Promise((resolve) => {
+            isProcessing = true;
+            function processNext() {
+                if (!isEmpty()) {
+                    const nextItem = dequeue();
+                    if (nextItem) {
+                        processItem(nextItem).then(processNext);
```

Comment:
What would happen if the processing fails?

Reply:
Also, I don't see a check enforcing a promise for a `processItem` function. Someone could pass in a `sync`. function and then this would get stuck forever.

Reply:
> What would happen if the processing fails?

It will be [canceled](https://github.com/Expensify/App/blob/36a80b5edede3ac0e36d50578a73bf3625438613/desktop/electronDownloadManager.ts#L70).

> Also, I don't see a check enforcing a promise for a processItem function. Someone could pass in a sync. function and then this would get stuck forever.

it's in the constructor

```ts
function createQueue<T>(processItem: (item: T) => Promise<void>): Queue<T> {
```

Reply:
> It will be [canceled](https://github.com/Expensify/App/blob/36a80b5edede3ac0e36d50578a73bf3625438613/desktop/electronDownloadManager.ts#L70).
Since we're making this a generic library, it makes sense to add a processing failure check inside this rather than leaving that to the user.

Reply:
@allroundexperts what should happen in case of failure? I believe it should continue to process the next item. If so, we can replace `then` with `finally`. Do you agree?

change:

```
processItem(nextItem).then(processNext);
```

To
```
processItem(nextItem).finally(processNext);
```

Reply:
We should log the error and move on.